syntax = "proto3";
package grpc.testing;

service Shengji {
  // Creates a new game.
  rpc createGame(CreateGameRequest) returns (Game);

  // Enters the room if the player has not entered the room yet. Streams back
  // the game whenever game state changes.
  rpc enterRoom(EnterRoomRequest) returns (stream Game);

  // Plays a hand or tries to claim trump suit.
  rpc playHand(PlayHandRequest) returns (PlayHandResponse);

  rpc addAIPlayer(AddAIPlayerRequest) returns (AddAIPlayerResponse);
}

// Request and response protos
message AddAIPlayerRequest {
  string game_id = 1;
}

message AddAIPlayerResponse {
  string player_name = 1;
}

message EnterRoomRequest {
  string game_id = 1;
  string player_id = 2;
}

message CreateGameRequest {
  string player_id = 1;
}

message PlayHandRequest{
  string player_id = 1;
  string game_id = 2;

  enum Intention {
    // Claims trump suit - this is only valid while cards are being served.
    CLAIM_TRUMP = 0;
    // Plays a hand.
    PLAY_HAND = 1;
    // Hides kitty cards.
    HIDE_KITTY = 2;
  }

  Intention intention = 3;
  Hand hand = 4;
}

message PlayHandResponse{
  // TODO: Ideally FE only needs to handle RPC failure. Use a boolean here for
  // now.
  bool success = 1;
  string error_message = 2;
}

// Player
message Player {
  // ID of the player
  string player_id = 1;

  // Cards you have on your hand
  Hand cards_on_hand = 2;

  // Cards you have in your winning pile
  Hand winning_pile = 3;

  // Cards played in the current ongoing round
  Hand current_round_trick = 4;

  // You current score
  int32 score = 5;
}

message Game {
  // Game ID.
  string game_id = 1;

  // Creator of the game.
  string creator_player_id = 2;

  // Dealer of this round. The dealer is the first player to play a hand and
  // has RO access to kitty.
  string dealer_player_id = 3;

  // Player who needs to play a hand.
  // Either "ANYONE" or a specific user id.
  //
  // ANYONE: Anyone can claim trump
  // specific user id: Only that user can play (hide kitty or play a hand)
  string next_turn_player_id = 4;

  // Player who wins the current round.
  string current_round_winner_player_id = 5;

  // All players. Private states will be redacted by the server (e.g.
  // other players' hand cards).
  // Order of players == play order.
  // The first player is guaranteed to be the caller.
  repeated Player players = 6;

  // Kitty/hidden hand. This is only visible to the dealer.
  Hand kitty = 7;

  // Current trump suit. If trump_suit is SUIT_UNDEFINED, then we're playing a
  // game without trump suit.
  Card.Suit trump_suit = 8;

  // Trump rank.
  Card.Rank trump_rank = 9;

  // Number of cards left in the deck. It should start with N (number of
  // decks) * 54 to start with, and become 0 once the dealing phase finishes.
  int32 deck_card_count = 10;

  oneof update {
    NewPlayerUpdate new_player_update = 11;
    CardDealtUpdate card_dealt_update = 12;
    KittyHiddenUpdate kitty_hidden_update = 13;
  }
}

message NewPlayerUpdate {
  // Player who joined.
  string player_id = 1;
}

message CardDealtUpdate {
  // Player to whom the card was dealt.
  string player_id = 1;
  Card card = 2;
}

message KittyHiddenUpdate {
  // Kitty player.
  string kitty_player_id = 1;
}

message Card {
  enum Suit {
    SUIT_UNDEFINED = 0;
    HEARTS = 1;
    SPADES = 2;
    CLUBS = 3;
    DIAMONDS = 4;

    SMALL_JOKER = 5;
    BIG_JOKER = 6;
  }

  enum Rank {
    RANK_UNDEFINED = 0;
    ACE = 1;
    TWO = 2;
    THREE = 3;
    FOUR = 4;
    FIVE = 5;
    SIX = 6;
    SEVEN = 7;
    EIGHT = 8;
    NINE = 9;
    TEN = 10;
    JACK = 11;
    QUEEN = 12;
    KING = 13;
  }

  Suit suit = 1;
  Rank rank = 2;
}

message Hand {
  repeated Card cards = 1;
}
